### DISCUSSION TITLE
"Storefront" to allow user customized screens without a code fork

### DISCUSSION THREAD

**Discussion by darianbjohnson** [Phambili-Tech/Newt_Display#35] [Ideas]
@badslug and I were [discussing ](https://github.com/Phambili-Tech/Newt_Display/discussions/20#discussioncomment-3530077)adding an way to access API data from his Tempest Weather System... which started us thinking about ways to add a "storefront" to allow users to add custom screens without recompiling code. Wanted to start a discussion on it and get some ideas on how to implement it.

**Comment by badslug** [comment]
A big design question would be whether users need/want multiple main screens or just a single one. For a single main screen, one potential approach would be a web app to customize the screen and "compile" some sort of data file that the Newt can download - either when attached via USB or maybe a download over the network. (Would that require an "account" of some sort?) Having the customization UI in a web app would make it much easier to implement. The compiling could even be generating a custom firmware that the user updates on their device (but how to integrate new changes into it). A potential model for that is keyboard firmware (see QMK or ZMK for examples): https://qmk.fm and their online configurator https://config.qmk.fm/

Flashing custom firmware is a bit clumsy/fiddly though.

ZMK https://zmk.dev takes a different approach where you fork the firmware repo which has github actions that will automatically build a new firmware as part of a commit (so the user customizes a config file and then commits it and github actions spits out a new firmware).

Flashing the firmware is still a last step that is still fiddly.

**Comment by angusmci** [comment]
I'm a big believer in the idea that creating a platform that supports plug-ins, extensions, etc. gives a product extra appeal. It caters to people's desire to fine-tune the system to their own needs, and it allows the platform to be used in ways that the original developer never imagined. So, +1 to this idea.

But it's complicated to implement, because you have to build the plug-in framework, build the 'store', and provide ways to get the plug-ins onto the device. It also has security implications: Newt malware may be constrained, but at the end of the day a Newt is still a general-purpose computer with a wi-fi connection, which gives it scope to cause quite a lot of trouble.

As far as the UI is concerned, maybe you could build on a 'widget' idea. Different elements of the core Newt firmware and user-added plug-ins could provide content for different parts of the homescreen. My own ideal homescreen display would probably look much like the current one, but I'd prefer to replace the guilt-inducing inspirational quotes with a list of my upcoming meetings or to-do items. Others will have their own preferences.

**Comment by tjfoth** [comment]
Also, I think we need to remember the limits of the Newt's storage and
processing power.

---
Tom Foth
***@***.***
***@***.***
